File: py/src/hop/aim/registry.py
"""Registry — async model registry backed by models.dev."""

from __future__ import annotations

from typing import Optional

from .query import parse_query
from .source import ModelsDevSource
from .types import Filter, Model, Provider

# NOTE: XDG cache is used in this Python port via platformdirs (optional dep).
# If platformdirs is not installed, in-memory cache is used instead.
# See testdata/xdg.md for XDG cache dir algorithm.

try:
    from platformdirs import user_cache_dir as _user_cache_dir
    _HAS_PLATFORMDIRS = True
except ImportError:
    _HAS_PLATFORMDIRS = False


def _xdg_cache_dir() -> Optional[str]:
    """Return XDG-compliant cache dir for hop/aim, or None if unavailable."""
    import os
    xdg = os.environ.get("XDG_CACHE_HOME")
    if xdg:
        return os.path.join(xdg, "hop", "aim")
    if _HAS_PLATFORMDIRS:
        return os.path.join(_user_cache_dir("hop", appauthor=False), "aim")
    return None


def _matches_filter(m: Model, f: Filter) -> bool:
    if f.input:
        for mod in f.input:
            if mod not in m.modalities.input:
                return False
    if f.output:
        for mod in f.output:
            if mod not in m.modalities.output:
                return False
    if f.provider and m.provider != f.provider:
        return False
    if f.family and m.family != f.family:
        return False
    if f.tool_call is not None and m.tool_call != f.tool_call:
        return False
    if f.reasoning is not None and m.reasoning != f.reasoning:
        return False
    if f.open_weights is not None and m.open_weights != f.open_weights:
        return False
    if f.structured_output is not None and m.structured_output != f.structured_output:
        return False
    if f.temperature is not None and m.temperature != f.temperature:
        return False
    if f.query:
        q = f.query.lower()
        if q not in m.id.lower() and q not in m.name.lower():
            return False
    return True


class Registry:
    """Async registry providing access to models.dev provider/model data."""

    def __init__(self, url: Optional[str] = None) -> None:
        self._source = ModelsDevSource(url=url) if url else ModelsDevSource()
        self._cache: Optional[dict[str, Provider]] = None

    async def refresh(self) -> None:
        """Force re-fetch from source."""
        self._cache = await self._source.fetch()

    async def _load(self) -> dict[str, Provider]:
        if self._cache is None:
            self._cache = await self._source.fetch()
        return self._cache

    async def providers(self) -> list[Provider]:
        data = await self._load()
        return list(data.values())

    async def models(self, filter: Optional[Filter] = None) -> list[Model]:
        data = await self._load()
        all_models: list[Model] = []
        for p in data.values():
            all_models.extend(p.models.values())
        if filter is None:
            return all_models
        return [m for m in all_models if _matches_filter(m, filter)]

    async def get(self, provider_id: str, model_id: str) -> Optional[Model]:
        data = await self._load()
        p = data.get(provider_id)
        if p is None:
            return None
        return p.models.get(model_id)

    async def query(self, q: str) -> list[Model]:
        f = parse_query(q)
        return await self.models(f)

File: ts/src/registry.ts
import type { Filter, Model, Provider } from './types.ts'
import { ModelsDevSource } from './source.ts'
import { parseQuery } from './query.ts'

// NOTE: XDG cache is skipped for the TS port — uses in-memory cache only
// (fetch once per Registry instance). See testdata/xdg.md for XDG algorithm
// used by Go and Python ports.

export interface RegistryOptions {
  url?: string
}

/** Registry provides access to the models.dev provider/model data. */
export class Registry {
  private readonly source: ModelsDevSource
  private cache: Record<string, Provider> | null = null

  constructor(opts: RegistryOptions = {}) {
    this.source = new ModelsDevSource({ url: opts.url })
  }

  /** refresh forces a re-fetch from the source. */
  async refresh(): Promise<void> {
    this.cache = await this.source.fetch()
  }

  private async load(): Promise<Record<string, Provider>> {
    if (!this.cache) {
      this.cache = await this.source.fetch()
    }
    return this.cache
  }

  /** providers returns all providers. */
  async providers(): Promise<Provider[]> {
    const data = await this.load()
    return Object.values(data)
  }

  /** models returns all models, optionally filtered. */
  async models(filter?: Filter): Promise<Model[]> {
    const data = await this.load()
    const all: Model[] = []
    for (const p of Object.values(data)) {
      for (const m of Object.values(p.models ?? {})) {
        if (m) all.push(m)
      }
    }
    if (!filter) return all
    return all.filter((m) => matchesFilter(m, filter))
  }

  /** get returns a single model by provider + model id. */
  async get(
    providerId: string,
    modelId: string,
  ): Promise<Model | undefined> {
    const data = await this.load()
    return data[providerId]?.models?.[modelId] ?? undefined
  }

  /** query parses the query string and returns matching models. */
  async query(q: string): Promise<Model[]> {
    const filter = parseQuery(q)
    return this.models(filter)
  }
}

function matchesFilter(m: Model, f: Filter): boolean {
  if (f.input?.length) {
    for (const mod of f.input) {
      if (!m.modalities.input.includes(mod)) return false
    }
  }
  if (f.output?.length) {
    for (const mod of f.output) {
      if (!m.modalities.output.includes(mod)) return false
    }
  }
  if (f.provider !== undefined && m.provider !== f.provider) return false
  if (f.family !== undefined && m.family !== f.family) return false
  if (f.toolCall !== undefined && m.tool_call !== f.toolCall) return false
  if (f.reasoning !== undefined && m.reasoning !== f.reasoning) return false
  if (f.openWeights !== undefined && m.open_weights !== f.openWeights) return false
  if (f.structuredOutput !== undefined && m.structured_output !== f.structuredOutput) return false
  if (f.temperature !== undefined && m.temperature !== f.temperature) return false
  if (f.query) {
    const q = f.query.toLowerCase()
    const idMatch = m.id.toLowerCase().includes(q)
    const nameMatch = m.name.toLowerCase().includes(q)
    if (!idMatch && !nameMatch) return false
  }
  return true
}
